src/pcg.rs

```diff
@@ -0,0 +1,157 @@
+// Copyright 2013 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! The Pcg family of random number generators.
+use std::num::Wrapping as w;
+
+use {Rng, SeedableRng, Rand, w32, w64};
+
+/// A PCG64[1] random number generator.
+///
+/// The PCG64 algorithm, more specifically PCG RXS M XS 64 (LCG),
+/// is not suitable for cryptographic purposes
+/// but is very fast. If you do not know for sure that it fits your
+/// requirements, use a more secure one such as `IsaacRng` or `OsRng`.
+///
+/// [1]: MELISSA E. O’NEILL (2015). [A PCG: A Family of Simple Fast Space-Efficient
+/// Statistically Good Algorithms for Random Number Generation.](http://www.pcg-random.org/paper.html)
+#[allow(missing_copy_implementations)]
+#[derive(Clone)]
+/// PCG RXS M XS 64 (LCG)
+pub struct Pcg64Rng {
+    state: w64
+}
+
+impl Pcg64Rng {
+    /// Creates a new Pcg64Rng instance which is not seeded.
+    ///
+    /// The initial values of this RNG are constants, so all generators created
+    /// by this function will yield the same stream of random numbers. It is
+    /// highly recommended that this is created through `SeedableRng` instead of
+    /// this function
+    pub fn new_unseeded() -> Pcg64Rng {
+        Pcg64Rng {
+            state: w(0x193a6754a8a7d469)
+        }
+    }
+}
+
+impl Rng for Pcg64Rng {
+    #[inline]
+    fn next_u32(&mut self) -> u32 {
+        self.next_u64() as u32
+    }
+
+    #[inline]
+    fn next_u64(&mut self) -> u64 {
+    	const MULTIPLIER: w64 = w(6364136223846793005);
+		const INCREMENT: w64  = w(1442695040888963407);
+    	let state = self.state;
+    	self.state = self.state * MULTIPLIER + INCREMENT;
+        let word = ((state >> (((state >> 59) + w(5)).0 as usize)) ^ state) * w(12605985483714917081);
+    	((word >> 43) ^ word).0
+    }
+}
+
+impl SeedableRng<u64> for Pcg64Rng {
+    /// Reseed an Pcg64Rng. This will panic if `seed` is entirely 0.
+    fn reseed(&mut self, seed: u64) {
+        assert!(seed == 0,
```

Comment:
Shouldn't this assert that the seed isn't 0? Can you be sure to add tests for all these implementations?